src/test/test_stats.c

```diff
@@ -245,6 +246,117 @@ test_rephist_mtbf(void *arg)
   tor_free(ddir_fname);
 }
 
+static void
+test_commit_max(void *arg)
+{
+  (void) arg;
+  bw_array_t *b = bw_array_new();
+  time_t now = b->cur_obs_time;
+
+  commit_max(b);
+  tt_int_op(b->next_period, OP_EQ, now + 2*86400);
+
+  b->total_in_period = 100;
+  b->max_total = 10;
+  commit_max(b);
+  tor_assert(b->total_in_period == 0);
+  tor_assert(b->max_total == 0);
+  tt_int_op(b->totals[1], OP_EQ, 100);
+  tt_int_op(b->maxima[1], OP_EQ, 10);
+  tt_int_op(b->next_period, OP_EQ, now + 3*86400);
+
+  commit_max(b);
+  tt_int_op(b->next_period, OP_EQ, now + 4*86400);
+
+  commit_max(b);
+  tt_int_op(b->next_period, OP_EQ, now + 5*86400);
+
+  b->total_in_period = 100;
+  b->max_total = 10;
+  commit_max(b);
+  tor_assert(!b->next_max_idx);
+  tt_int_op(b->cur_obs_idx, OP_EQ, 0);
+  tt_int_op(b->totals[4], OP_EQ, 100);
+  tt_int_op(b->maxima[4], OP_EQ, 10);
+  tt_int_op(b->next_period, OP_EQ, now + 6*86400);
+ done:
+  bw_array_free(b);
+}
+
+#define test_obs(b, idx, time, tot, max) STMT_BEGIN \
+    tt_int_op(b->cur_obs_idx, OP_EQ, idx); \
+    tt_int_op(b->cur_obs_time, OP_EQ, time); \
+    tt_int_op(b->total_obs, OP_EQ, tot); \
+    tt_int_op(b->max_total, OP_EQ, max); \
+  STMT_END;
+
+static void
+test_advance_obs(void *arg)
+{
+  (void) arg;
+  int iter, tot = 0;
+  bw_array_t *b = bw_array_new();
+  time_t now = b->cur_obs_time;
+
+  for (iter = 0; iter < 10; ++iter) {
+    b->obs[b->cur_obs_idx] += 10;
+    tot += 10;
+    advance_obs(b);
+    if (iter == 9) {
+      /* The current value under cur_obs_idx was zeroed in last iterN. */
+      test_obs(b, 0, now+iter+1, tot - 10, tot);
+      break;
+    }
+    test_obs(b, iter+1, now+iter+1, tot, tot);
+  }
+
+  b->total_in_period = 100;
+  b->cur_obs_time = now + NUM_SECS_BW_SUM_INTERVAL - 1;
+  advance_obs(b);
+  test_obs(b, 1, now+NUM_SECS_BW_SUM_INTERVAL, 80, 0);
+  tt_int_op(b->maxima[0], OP_EQ, 100);
+  tt_int_op(b->totals[0], OP_EQ, 100);
+  tt_int_op(b->num_maxes_set, OP_EQ, 1);
+ done:
+  bw_array_free(b);
+}
+
+#define test_add_obs_(b, now, checknow, bw, tot) STMT_BEGIN \
+    tot += bw; \
+    add_obs(b, now, bw); \
+    tt_int_op(b->cur_obs_time, OP_EQ, checknow); \
+    tt_int_op(b->obs[b->cur_obs_idx], OP_EQ, bw); \
+    tt_int_op(b->total_in_period, OP_EQ, tot); \
+  STMT_END;
+
+static void
+test_add_obs(void *arg)
+{
+  (void) arg;
+  bw_array_t *b = bw_array_new();
+  time_t now = b->cur_obs_time;
+  uint64_t bw = 0, tot = 0;
+  /* Requests for the past should not be entertained. */
+  test_add_obs_(b, now-1, now, bw, tot);
+  /* Test the expected functionalities for random values. */
+  now += 53;
+  bw = 97;
+  test_add_obs_(b, now, now, bw, tot);
+
+  now += 60*60;
+  bw = 90;
+  test_add_obs_(b, now, now, bw, tot);
+
+  /* We should not test for new period because the time complexity
+   * will be very high. */
```

Comment:
Can you explain what you mean by "the time complexity will be very high" ?

Reply:
The loop [here](https://github.com/torproject/tor/blob/686494f0f71b9235399b8241aba3e0c2fcb03ea1/src/feature/stats/rephist.c#L1074) will iterate
86400 times for a day.
And you can see the comment below 
describing this behavior.
Should I test it for a day forward, nevertheless ??

Reply:
How slow is the test when you run it?

80,000 is not much, for a computer that runs a billion operations per second.